plugins/pulp_docker/plugins/importers/sync.py

```diff
@@ -300,3 +300,51 @@ def process_main(self):
                                                       tag_name=tag, manifest_digest=manifest.digest)
             if new_tag:
                 repository.associate_single_unit(self.get_repo().repo_obj, new_tag)
+
+
+class TokenAuthDownloadStep(publish_step.DownloadStep):
+    """
+    Download remote files. For v2, this may require a bearer token to be used. This step attempts
+    to download files, and if it fails due to a 401, it will retrieve the auth token and retry the
+    download.
+    """
+
+    def __init__(self, step_type, downloads=None, repo=None, conduit=None, config=None,
+                 working_dir=None, plugin_type=None, description=''):
+        """
+        Initialize the step, setting its description.
+        """
+
+        super(TokenAuthDownloadStep, self).__init__(
+            step_type, downloads=downloads, repo=repo, conduit=conduit, config=config,
+            working_dir=working_dir, plugin_type=plugin_type)
+        self.description = _('Downloading remote files')
+        self.token = None
+
+    def process_main(self, item=None):
+        """
+        Overrides the parent method to get a new token and try again if response is a 401.
+        """
+        # Allow the original request to be retrieved from the url.
+        self._requests_map = {}
```

Comment:
Some would say (including at least one static analysis tool I've seen- maybe a optional flake8 plugin?) that this should be done in ``__init__``. In general I think that's a good idea, and I don't see a compelling reason here to not do so.